File: src/org/kobjects/nlp/latin/Latin.java
package org.kobjects.nlp.latin;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;

import org.kobjects.nlp.api.Definition;
import org.kobjects.nlp.api.Form;
import org.kobjects.nlp.api.Case;
import org.kobjects.nlp.api.Gender;
import org.kobjects.nlp.api.Language;
import org.kobjects.nlp.api.Number;
import org.kobjects.nlp.api.Word;
import org.kobjects.nlp.api.WordType;

public class Latin implements Language {

	static final String VOCALS = "aeiouAEIOU";
	static final String CONSONANTS = "bcdfghjklmnpqrstvwxyzBCDFGHJKLMNPQRSTVWXYZ";

	static final Case[] CASES = { 
			Case.NOMINATIVE, 
			Case.DATIVE, 
			Case.GENITIVE, 
			Case.ACCUSATIVE, 
			Case.ABLATIVE,
			Case.VOCATIVE
	};

	static final Number[] NUMERI = { Number.SINGULAR, Number.PLURAL };

	static final Gender[] GENERA = { Gender.MASCULINE, Gender.FEMININE, Gender.NEUTER };

	static boolean isConsonant(char c) {
		return CONSONANTS.indexOf(c) != -1;
	}

	static boolean isVocal(char c) {
		return VOCALS.indexOf(c) != -1;
	}

	/**
	 * Checks whether all characters in s are consonants.
	 */
	static boolean consonantsOnly(String s) {
		for (int i = 0; i < s.length(); i++) {
			if (CONSONANTS.indexOf(s.charAt(i)) == -1) {
				return false;
			}
		}
		return true;
	}

	static int syllableCount(String s) {
		int count = 0;
		for (int i = 0; i < s.length(); i++) {
			if (VOCALS.indexOf(s.charAt(i)) != -1) {
				count++;
			}
		}
		return count;
	}

	Map<String, Set<Word>> dictionary = new TreeMap<>();

	public Latin() throws IOException {
		BufferedReader reader = new BufferedReader(
				new InputStreamReader(getClass().getResourceAsStream("whitaker_converted.txt"), "utf-8"));

		String nextLine = reader.readLine();
		while (nextLine != null) {
			String line = nextLine.trim();
			nextLine = reader.readLine();
			while (nextLine != null && nextLine.startsWith(" ")) {
				line += ' ' + nextLine.trim();
				nextLine = reader.readLine();
			}
			try {
				Definition definition = parseDefinition(line);
				if (definition != null) {
					for (Word word : definition.forms.values()) {
						String s = word.word;
						Set<Word> words = dictionary.get(s);
						if (words == null) {
							words = new LinkedHashSet<Word>();
							dictionary.put(s, words);
						}
						words.add(word);
					}
				}
			} catch (Exception e) {
				System.err.println("Error processing '" + line + "': " + e.getMessage());
				e.printStackTrace();
			}
		}
	}

	public Definition parseDefinition(String def) {
		def = def.toLowerCase();
		int colon = def.indexOf(":");
		if (colon == -1) {
			throw new RuntimeException(": expected.");
		}
		
		String[] leftParts = def.substring(0, colon).split(";");
		String[] wordParts = leftParts[0].trim().split(",");
		String[] typeParts = leftParts[1].trim().split(" ");
   	 	String[] metaParts = leftParts.length > 2 ? leftParts[2].split(",") : new String[0];

   	 	for (String meta: metaParts) {
   	 		meta = meta.trim();
   	 		if (meta.startsWith("f:") && meta.indexOf("frequent") == -1 && !meta.equals("f:common")) {
   	 			return null;
   	 		}
   	 	}

   	 	for (int i = 0; i < wordParts.length; i++) {
   	 		String word = wordParts[i].trim();
   	 		if (word.endsWith("(i)") || word.endsWith("(ii)")) {
   	 			word = word.substring(0, word.indexOf('(')) + 'i';
   	 		}
   	 		wordParts[i] = word;
   	 	}
   	 	
		String wordTypeName = typeParts[0];
		WordType type = null;
		for (WordType t : WordType.values()) {
			if (t.toString().equals(wordTypeName.toUpperCase())) {
				type = t;
				break;
			}
		}
		if (type == null) {
			throw new RuntimeException("Unrecognized word type: " + wordTypeName);
		}
		Definition definition = new Definition(type, def.substring(colon + 2).trim());
		
		switch (type) {
		case NOUN:
			processNoun(definition, wordParts, typeParts);
			break;
		case VERB:
			processVerb(definition, wordParts, typeParts);
			break;
		case ADJECTIVE:
			definition.addForms(Declinator.declineAdjective(wordParts));
			break;
		default:
			definition.addFormless(wordParts[0]);
		}
		return definition;
	}

	/**
	 * Bildet alle Formen eines Substantivs.
	 */
	private void processNoun(Definition definition, String[] word, String[] kind) {
		switch (kind[kind.length - 1].toUpperCase().trim()) {
		case "F":
			definition.genus = Gender.FEMININE;
			break;
		case "C":
		case "M":
			definition.genus = Gender.MASCULINE;
			break;
		case "N":
			definition.genus = Gender.NEUTER;
			break;
		default:
			throw new RuntimeException("Unrecognized genus '" + kind[kind.length - 1] + "'");
		}
		if (word.length != 2) {
		  throw new RuntimeException("Nominative and genitive expected for noun entries. Got: " + Arrays.toString(word));
		}
		definition.addForms(Declinator.declineNoun(definition.genus, word[0], word[1]));
	}

	/**
	 * Bildet alle Formen eines Verbs.
	 */
	void processVerb(Definition definition, String[] word, String[] kind) {
		String conjugationDescription = kind.length > 1 ? kind[1] : "0";
		if (conjugationDescription.startsWith("(")) {
			conjugationDescription = conjugationDescription.substring(1, 2);
		}
		int conjugation = conjugationDescription.length() == 1 && conjugationDescription.charAt(0) >= '1'
				&& conjugationDescription.charAt(0) <= '4' ? Integer.parseInt(conjugationDescription) : 0;

		for (Map.Entry<Form, String> entry : Conjugator.conjugate(conjugation, word).entrySet()) {
			definition.forms.put(entry.getKey(), new Word(entry.getValue(), entry.getKey(), definition));
		}
	};

	public String getName() {
		return "Latin";
	}

	@Override
	public Set<Word> find(String word) {
		return dictionary.get(word);
	}

  public static String getStem(String nominative) {
    if (isVocal(nominative.charAt(nominative.length() - 1))) {
      return nominative.substring(0, nominative.length() - 1);
    }
    return nominative.substring(0, nominative.length() - 2);
  }

}

File: src/latin/Latin.js
import Case from "../api/Case.js";
import Definition from "../api/Definition.js";
import Gender from "../api/Gender.js";
import Plurality from "../api/Plurality.js";
import WordType from "../api/WordType.js";
import Conjugation from "./Conjugation.js";
import conjugate from "./Conjugator.js";
import Declinator from "./Declinator.js";

export default class Latin {
  static VOCALS = "aeiouAEIOU";
  static CONSONANTS = "bcdfghjklmnpqrstvwxyzBCDFGHJKLMNPQRSTVWXYZ";

  static CASES = Object.values(Case);

  static PLURALITY = Object.values(Plurality);

  static GENERA = Object.values(Gender);

  static isConsonant(char) {
    return this.CONSONANTS.indexOf(char) !== -1;
  }

  static isVocal(char) {
    return this.VOCALS.indexOf(char) !== -1;
  }

  /**
   * Checks whether all characters in s are consonants.
   */
  static consonantsOnly(string) {
    for (let i = 0; i < string.length; i++) {
      if (this.CONSONANTS.indexOf(string.charAt(i)) === -1) {
        return false;
      }
    }
    return true;
  }

  static findStem(nominative) {
    const length = nominative.length;
    if (this.isVocal(nominative.charAt(length - 1))) {
      return nominative.substring(0, length - 1);
    }
    return nominative.substring(0, length - 2);
  }

  static syllableCount(string) {
    let count = 0;
    for (let i = 0; i < string.length; i++) {
      if (this.VOCALS.indexOf(string.charAt(i)) > -1) {
        count++;
      }
    }
    return count;
  }

  constructor(rawText) {
    this.dictionary = new Map();

    // TODO: Use Stream if .txt file is still used
    const lines = rawText.split("\n");
    const rawDefinitions = lines.reduce((accumulator, current, index) => {
      // A definition
      if (current.startsWith(" ")) {
        accumulator[accumulator.length - 1] += ` ${current.trim()}`
      // The main line
      } else {
        accumulator.push(current);
      }
      return accumulator;
    }, [])

    let errors = 0;
		
    rawDefinitions.forEach(rawLine => {
      let line = rawLine.trim();
      try {
        const definition = this.parseDefinition(line);
        if (definition == null) {
          return;
        }
        for (let entry of definition.forms) {
          const word = entry[1];
          const string = word.word;
          let words = this.dictionary.get(string);
          if (words == null) {
            words = new Set();
            this.dictionary.set(string, words);
          }
          words.add(word);
        }
      } catch (err) {
        errors++;
        console.error("\n\nError processing '" + line + "': " + err + "\n");
        console.error(err.stack);
      }  
    });
    console.error("\nCount: " + rawDefinitions.length + "\nErrors: " + errors + "\nSuccessful: " + (rawDefinitions.length - errors));
  }

  parseDefinition(def) {
    def = def.toLowerCase();
    const colon = def.indexOf(":");
    if (colon === -1) {
      throw new Error(": expected.\nDef: " + def);
    }
    
    const leftParts = def.substring(0, colon).split(";");
    const wordParts = leftParts[0].trim().split(", ");
    const typeParts = leftParts[1].trim().split(" ");
    /*    
      TODO: (Looks like something unfinished from Java project)

      String[] metaParts = leftParts.length > 2 ? leftParts[2].split(",") : new String[0];
      Skip infrequent
        for (String meta: metaParts) {
    
          meta = meta.trim();
          if (meta.startsWith("frq:") && meta.indexOf("-") !== -1) {
            return null;
          }
        }
    */
    
    for (let i = 0; i < wordParts.length; i++) {
      let word = wordParts[i].trim();
      if (word.endsWith("(i)") || word.endsWith("(ii)")) {
        word = word.substring(0, word.indexOf('(')) + 'i';
      }
      wordParts[i] = word;
    }
    
    const wordTypeName = typeParts[0];
    let type = null;
    for (let t of Object.values(WordType)) {
      if (t.toString() === wordTypeName.toUpperCase()) {
        type = t;
        break;
      }
    }
    if (type == null) {
      throw new Error("Unrecognized word type: " + wordTypeName);
    }

    const definition = new Definition(type, def.substring(colon + 2).trim());
    switch (type) {
      case WordType.NOUN:
        return this.processNoun(definition, wordParts, typeParts);
      case WordType.VERB:
        return this.processVerb(definition, wordParts, typeParts);
      case WordType.ADJECTIVE:
        return definition.addForms(Declinator.declineAdjective(wordParts));
      default:
        return definition.addFormless(wordParts[0]);
    }
  }

  /**
   * All forms of a noun
   */
  processNoun(definition, words, kinds) {
    switch (kinds[kinds.length - 1].toUpperCase().trim()) {
      case "F":
        definition.genus = Gender.FEMININE;
        break;
      case "C":
      case "M":
        definition.genus = Gender.MASCULINE;
        break;
      case "N":
        definition.genus = Gender.NEUTER;
        break;
      default:
        throw new Error("Unrecognized genus '" + kinds[kinds.length - 1] + "'");
    }
    if (words.length !== 2) {
      throw new Error("Nominative and genitive expected for noun entries. Got: " + words.join(", "));
    }
    const declinedNoun = Declinator.declineNoun(definition.genus, words[0], words[1]);
    return definition.addForms(declinedNoun);
  }

  /**
   * All forms of a verb
   */
  processVerb(definition, words, kinds) {
    let conjugationDescription = kinds.length > 1 ? kinds[1] : "0";
    if (conjugationDescription.startsWith("(")) {
      conjugationDescription = conjugationDescription.substring(1, 2);
    }
    const hasConjugation = (
      conjugationDescription.length === 1 &&
      conjugationDescription.charAt(0) >= '1' &&
      conjugationDescription.charAt(0) <= '4'
    ) 
    const conjugation = hasConjugation ? parseInt(conjugationDescription) : 0;
    return definition.addForms(conjugate(words[0], words[1], words[2], words[3], conjugation));
  }

  find(word) {
    return this.dictionary.get(word);
  }
}
